src/tests.rs

```diff
@@ -1456,6 +1480,491 @@ fn test_secret_redaction() {
+                    \"error_description\": \"Access Denied\"\
+                    }"
+                .to_string()
+                .into_bytes(),
+            },
+        ),
+        mock_sleep_fn,
+        None)
+        .err()
+        .unwrap();
+    match token {
+        RequestTokenError::ServerResponse(msg) => {
+            assert_eq!(msg.error(), &DeviceCodeErrorResponseType::AccessDenied)
+        }
+        _ => unreachable!("Error should be Access Denied"),
+    }
+}
+
+#[test]
+fn test_device_token_expired() {
+    let details = new_device_auth_details(2);
+    assert_eq!("12345", details.device_code().secret());
+    assert_eq!("https://verify/here", details.verification_uri().as_str());
+    assert_eq!("abcde", details.user_code().secret().as_str());
+    assert_eq!(
+        "https://verify/here?abcde",
+        details
+            .verification_uri_complete()
+            .unwrap()
+            .secret()
+            .as_str()
+    );
+    assert_eq!(Duration::from_secs(2), details.expires_in());
+    assert_eq!(Duration::from_secs(1), details.interval());
+
+    let token = new_client()
+        .exchange_device_access_token(&details)
+        .set_time_fn(mock_time_fn())
+        .request(mock_http_client(
+            vec![
+                (ACCEPT, "application/json"),
+                (CONTENT_TYPE, "application/x-www-form-urlencoded"),
+                (AUTHORIZATION, "Basic YWFhOmJiYg=="),
+            ],
+            "grant_type=urn%3Aietf%3Aparams%3Aoauth%3Agrant-type%3Adevice_code&device_code=12345",
+            None,
+            HttpResponse {
+                status_code: StatusCode::from_u16(400).unwrap(),
+                headers: vec![(
+                    CONTENT_TYPE,
+                    HeaderValue::from_str("application/json").unwrap(),
+                )]
+                .into_iter()
+                .collect(),
+                body: "{\
+                    \"error\": \"expired_token\", \
+                    \"error_description\": \"Token has expired\"\
+                    }"
+                .to_string()
+                .into_bytes(),
+            },
+        ),
+        mock_sleep_fn,
+        None)
+        .err()
+        .unwrap();
+    match token {
+        RequestTokenError::ServerResponse(msg) => {
+            assert_eq!(msg.error(), &DeviceCodeErrorResponseType::ExpiredToken)
+        }
+        _ => unreachable!("Error should be ExpiredToken"),
+    }
+}
+
+fn mock_http_client_success_fail(
+    request_url: Option<Url>,
+    request_headers: Vec<(HeaderName, &'static str)>,
+    request_body: &'static str,
+    failure_response: HttpResponse,
+    num_failures: usize,
+    success_response: HttpResponse,
+) -> impl Fn(HttpRequest) -> Result<HttpResponse, FakeError> {
+    let mut responses: Vec<HttpResponse> = std::iter::repeat(failure_response)
+        .take(num_failures)
+        .collect();
+    responses.push(success_response);
```

Comment:
nit: not that it really matters for a test, but I think it's slightly more efficient/idiomatic to do `.chain(std::iter::once(success_response))` prior to `.collect`

Reply:
Done